R/layout.R

```diff
@@ -108,3 +112,312 @@ layout_column_wrap <- function(
   )
 }
 
+#' Bootstrap grid layout
+#'
+#' TODO: describe me
+#'
+#' @export
+layout_columns <- function(
+  ...,
+  widths = NA,
+  gap = "1rem",
+  fill = TRUE,
+  fillable = TRUE,
+  height = NULL,
+  height_mobile = NULL,
+  class = NULL,
+  n_cols = NULL
+) {
+  # TODO: should we catch `width` vs `widths`?
+
+  args <- list_split_named(rlang::list2(...))
+  attribs <- args[["named"]]
+  children <- dropNulls(args[["unnamed"]])
+  n_kids <- length(children)
+
+  # If no widths info is supplied, define number of columns as the number of elements
+  if (is.null(n_cols)) {
+    n_cols <- if (all(is.na(widths))) n_kids else 12
+  }
+
+  if (!is_breakpoints(widths, "columns")) {
+    widths <- breakpoints_columns(md = widths)
+  }
+
+  width_classes <- bs_css_grid_width_classes(widths, n_kids, n_cols)
+
+  children <- Map(f = bs_grid_wrapper, children, width_classes, fillable)
+
+  tag <- div(
+    class = "grid bslib-grid",
+    style = css(
+      "--bs-gap" = gap,
+      "--bs-columns" = n_cols,
+      # Always provide the `height:auto` default so that the CSS variable
+      # doesn't get inherited in a scenario like layout_column_wrap(height=200, ..., layout_column_wrap(...))
+      "--bslib-grid-height" = validateCssUnit(height %||% "auto"),
+      "--bslib-grid-height-mobile" = validateCssUnit(height_mobile %||% "auto"),
+      # For some reason, Bootstrap sets `grid-template-rows: 1fr` by default, which is
+      # problematic for a multi-row filling layout. This fixes it...
+      # > page_fillable(layout_columns(c(12, 12), plotly::plot_ly(), plotly::plot_ly()))
+      grid_auto_rows = "1fr"
+    ),
+    !!!attribs,
+    !!!children
+  )
+
+  tag <- bindFillRole(tag, item = fill)
+  tag <- tagAppendAttributes(tag, class = class)
+
+  as_fragment(
+    tag_require(tag, version = 5, caller = "layout_column_grid()")
+  )
+}
+
+bs_grid_wrapper <- function(el, bs_grid_classes = NULL, fillable = TRUE) {
+  div(
+    class = bs_grid_classes,
+    class = "bslib-gap-spacing",
+    if (fillable) as_fillable_container(),
+    el
+  )
+}
+
+bs_css_grid_width_classes <- function(breakpoints, n_kids, n_cols = 12) {
+  base_cols <- if (is_breakpoints(n_cols)) 12 else n_cols
+  classes <- as.list(rep(paste0("g-col-", base_cols), n_kids))
+
+  add_class <- function(idx, new) {
+    classes[[idx]] <<- c(classes[[idx]], new)
+  }
+
+  for (break_name in names(breakpoints)) {
+    bk <- breakpoints[[break_name]]
+    bk_cols <-
+      if (inherits(n_cols, "bslib_breakpoints")) {
+        n_cols[[break_name]]
+      } else {
+        n_cols
+      }
+
+    if (is.null(bk_cols)) {
```

Comment:
Let's add a comment about how the `NULL` case can happen

Reply:
I'm going to roll back allowing `n_cols` to be given breakpoints. Mostly because I realized we'll have to take a different approach. This function currently assumes that breakpoints are driven by widths; this NULL case happened when there was a mismatch between width and `n_cols` break points. But we'd want to do the same for `n_cols`, e.g. widths are `c(2, 2)` at `md` and `n_cols` is `breakpoints(md = 2, lg = 4, xl = 6)`.